# **Trabajo Práctico - Reglas de Asociación - Lisandro Duhalde**


En este trabajo se aplican reglas de asociación para el análisis de patrones frecuentes en un conjunto de datos transaccionales.  
Se utiliza el algoritmo Apriori para identificar itemsets frecuentes y extraer reglas de asociación relevantes, evaluándolas mediante métricas como soporte, confianza y lift.

El objetivo es descubrir relaciones significativas entre los elementos del dataset y analizar su interpretación desde el punto de vista del negocio/dominio del problema.



## Técnicas utilizadas
- Reglas de asociación
- Algoritmo Apriori
- Soporte, confianza y lift

In [ ]:
import warnings

# Ignorar DeprecationWarning del modulo jupyter_client
warnings.filterwarnings("ignore", category=DeprecationWarning, module="jupyter_client")

Cargamos la librería que nos permite acceder al dataset

In [ ]:
# soporte para cargar dataset de https://www.openml.org/
!pip install openml
import openml

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 160.4/160.4 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 93.1/93.1 kB 3.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 27.0 MB/s eta 0:00:00
  Created wheel for liac-arff: filename=liac_arff-2.5.0-py3-none-any.whl size=11717 sha256=c936be80e3645fbdbfc17e4192bdc7736cba17ba6ba845b2829ad4e24525a2bd
  Stored in directory: /root/.cache/pip/wheels/a9/ac/cf/c2919807a5c623926d217c0a18eb5b457e5c19d242c3b5963a
Successfully built liac-arff


Accedemos al dataset

In [ ]:
import pandas as pd

# indicamos cual dataset queremos utilizar, en este caso el nro. 24
dataset = openml.datasets.get_dataset(42585)

# separamos las información almacenada en el dataset
X, y, categorical_indicator, attribute_names = dataset.get_data(
    dataset_format='dataframe',
    target=dataset.default_target_attribute
)

#  concatenamos la información relevante en un único DataFrame
df = pd.concat([X, y], axis=1)
df

island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0    Torgersen              39.1             18.7              181.0   
1    Torgersen              39.5             17.4              186.0   
2    Torgersen              40.3             18.0              195.0   
3    Torgersen               NaN              NaN                NaN   
4    Torgersen              36.7             19.3              193.0   
..         ...               ...              ...                ...   
339     Biscoe               NaN              NaN                NaN   
340     Biscoe              46.8             14.3              215.0   
341     Biscoe              50.4             15.7              222.0   
342     Biscoe              45.2             14.8              212.0   
343     Biscoe              49.9             16.1              213.0   

     body_mass_g     sex species  
0         3750.0    MALE  Adelie  
1         3800.0  FEMALE  Adelie  
2         3250.0  FEMALE  Adelie  
3            NaN     NaN  Adelie  
4         3450.0  FEMALE  Adelie  
..           ...     ...     ...  
339          NaN     NaN  Gentoo  
340       4850.0  FEMALE  Gentoo  
341       5750.0    MALE  Gentoo  
342       5200.0  FEMALE  Gentoo  
343       5400.0    MALE  Gentoo  

[344 rows x 7 columns]

a) El dataset seleccionado contiene información sobre pingüinos de la base Palmer, un laboratorio en la Antártida. Sobre un total de 344 pingüinos, se los clasifica de acuerdo con 7 características.

Dentro de las 7 características, la especie, la isla y el sexo son de tipo nominal, y solo pueden tomar 3 valores. El resto son de tipo numérico y pueden tomar un rango mayor de valores.

b) Aplicar las reglas de asociación en este dataset puede ser útil para encontrar distintos tipos de relaciones entre las variables estudiadas, como por ejemplo si dentro de alguna isla predomina alguna especie, o cómo varía el peso de los pingüinos por especie/isla, también cómo varían las medidas de los pingüinos de distinta especie/género, etc.

c) Para las filas con valores faltantes, se eliminaran esas filas, ya que para discretizar los intervalos es necesario remover los valores faltantes; para la variable 'sex', se reemplazaron las celdas con valor '_' por NaN para que las elimine correctamente la funcion 'dropna'.

Luego vamos a discretizar en 3 intervalos a las variables continuas.

Y por ultimo se utilizara One-Hod Encoding para que algoritmo Apiori funcione correctamente.


In [ ]:
import pandas as pd
from sklearn.preprocessing import KBinsDiscretizer

df['sex'] = df['sex'].replace('_', pd.NA)

columnas_cat = ["island","sex","species"]
df_limpio = df.dropna(subset=columnas_cat)

columnas_cont = ["culmen_length_mm", "culmen_depth_mm", "flipper_length_mm", "body_mass_g"]
kbd = KBinsDiscretizer(n_bins=3, encode="ordinal", strategy="quantile")
df_limpio[columnas_cont] = kbd.fit_transform(df_limpio[columnas_cont])

df_procesado = pd.get_dummies(df_limpio, columns=columnas_cat + columnas_cont)
df_procesado

/tmp/ipython-input-1679303345.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_limpio[columnas_cont] = kbd.fit_transform(df_limpio[columnas_cont])


island_Biscoe  island_Dream  island_Torgersen  sex_FEMALE  sex_MALE  \
0            False         False              True       False      True   
1            False         False              True        True     False   
2            False         False              True        True     False   
4            False         False              True        True     False   
5            False         False              True       False      True   
..             ...           ...               ...         ...       ...   
338           True         False             False        True     False   
340           True         False             False        True     False   
341           True         False             False       False      True   
342           True         False             False        True     False   
343           True         False             False       False      True   

     species_Adelie  species_Chinstrap  species_Gentoo  culmen_length_mm_0.0  \
0              True              False           False                  True   
1              True              False           False                  True   
2              True              False           False                  True   
4              True              False           False                  True   
5              True              False           False                  True   
..              ...                ...             ...                   ...   
338           False              False            True                 False   
340           False              False            True                 False   
341           False              False            True                 False   
342           False              False            True                 False   
343           False              False            True                 False   

     culmen_length_mm_1.0  culmen_length_mm_2.0  culmen_depth_mm_0.0  \
0                   False                 False                False   
1                   False                 False                False   
2                   False                 False                False   
4                   False                 False                False   
5                   False                 False                False   
..                    ...                   ...                  ...   
338                 False                  True                 True   
340                 False                  True                 True   
341                 False                  True                 True   
342                  True                 False                 True   
343                 False                  True                 True   

     culmen_depth_mm_1.0  culmen_depth_mm_2.0  flipper_length_mm_0.0  \
0                  False                 True                   True   
1                   True                False                   True   
2                   True                False                  False   
4                  False                 True                  False   
5                  False                 True                   True   
..                   ...                  ...                    ...   
338                False                False                  False   
340                False                False                  False   
341                False                False                  False   
342                False                False                  False   
343                False                False                  False   

     flipper_length_mm_1.0  flipper_length_mm_2.0  body_mass_g_0.0  \
0                    False                  False            False   
1                    False                  False            False   
2                     True                  False             True   
4                     True                  False             True   
5                    False                  False  

d) Para obtener las reglas de asociación, primero obtenemos los itemsets frecuentes a partir del algoritmo Apriori y luego generamos las reglas de asociación a partir de esos itemsets.El soporte (0.15) y la confianza (0.75) poseen valores moderados, ya que luego en la etapa de post procesamiento se van a filtrar un poco mas las reglas irrelelevantes.

In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = apriori(df_procesado, min_support=0.15,use_colnames=True)

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.75)

rules

antecedents  \
0                                     (species_Gentoo)   
1                                (culmen_depth_mm_0.0)   
2                              (flipper_length_mm_2.0)   
3                                    (body_mass_g_2.0)   
4                                  (species_Chinstrap)   
..                                                 ...   
617  (culmen_length_mm_2.0, culmen_depth_mm_0.0, bo...   
618  (culmen_length_mm_2.0, flipper_length_mm_2.0, ...   
619              (island_Biscoe, culmen_length_mm_2.0)   
620             (culmen_length_mm_2.0, species_Gentoo)   
621        (culmen_length_mm_2.0, culmen_depth_mm_0.0)   

                                           consequents  antecedent support  \
0                                      (island_Biscoe)            0.357357   
1                                      (island_Biscoe)            0.327327   
2                                      (island_Biscoe)            0.345345   
3                                      (island_Biscoe)            0.336336   
4                                       (island_Dream)            0.204204   
..                                                 ...                 ...   
617  (island_Biscoe, species_Gentoo, flipper_length...            0.159159   
618   (island_Biscoe, species_Gentoo, body_mass_g_2.0)            0.156156   
619  (culmen_depth_mm_0.0, flipper_length_mm_2.0, s...            0.201201   
620  (culmen_depth_mm_0.0, island_Biscoe, flipper_l...            0.201201   
621  (flipper_length_mm_2.0, island_Biscoe, species...            0.162162   

     consequent support   support  confidence      lift  representativity  \
0              0.489489  0.357357    1.000000  2.042945               1.0   
1              0.489489  0.315315    0.963303  1.967974               1.0   
2              0.489489  0.330330    0.956522  1.954121               1.0   
3              0.489489  0.318318    0.946429  1.933501               1.0   
4              0.369369  0.204204    1.000000  2.707317               1.0   
..                  ...       ...         ...       ...               ...   
617            0.330330  0.153153    0.962264  2.913036               1.0   
618            0.309309  0.153153    0.980769  3.170836               1.0   
619            0.252252  0.153153    0.761194  3.017591               1.0   
620            0.252252  0.153153    0.761194  3.017591               1.0   
621            0.297297  0.153153    0.944444  3.176768               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0    0.182435         inf       0.794393  0.730061   1.000000    0.865031  
1    0.155092   13.911411       0.731207  0.628743   0.928117    0.803737  
2    0.161287   11.741742       0.745830  0.654762   0.914834    0.815684  
3    0.153685    9.529530       0.727482  0.627219   0.895063    0.798368  
4    0.128777         inf       0.792453  0.552846   1.000000    0.776423  
..        ...         ...            ...       ...        ...         ...  
617  0.100578   17.746246       0.781022  0.455357   0.943650    0.712950  
618  0.104853   35.915916       0.811318  0.490385   0.972157    0.737957  
619  0.102400    3.131194       0.837019  0.510000   0.680633    0.684168  
620  0.102400    3.131194       0.837019  0.510000   0.680633    0.684168  
621  0.104943   12.648649       0.817837  0.500000   0.920940    0.729798  

[622 rows x 14 columns]

e) Para la etapa de post procesamiento, ademas de haber utilizado soporte y confianza previamente, vamos a utilizar valores altos para lift (2) y conviction (3), para filtrar reglas que no sean utiles.

In [ ]:
rules_filtradas = rules[(rules['conviction'] > 3) & (rules['lift'] > 2)]

rules_filtradas



antecedents  \
0                                     (species_Gentoo)   
4                                  (species_Chinstrap)   
8                               (culmen_length_mm_0.0)   
11                                    (species_Gentoo)   
12                               (culmen_depth_mm_0.0)   
..                                                 ...   
617  (culmen_length_mm_2.0, culmen_depth_mm_0.0, bo...   
618  (culmen_length_mm_2.0, flipper_length_mm_2.0, ...   
619              (island_Biscoe, culmen_length_mm_2.0)   
620             (culmen_length_mm_2.0, species_Gentoo)   
621        (culmen_length_mm_2.0, culmen_depth_mm_0.0)   

                                           consequents  antecedent support  \
0                                      (island_Biscoe)            0.357357   
4                                       (island_Dream)            0.204204   
8                                     (species_Adelie)            0.333333   
11                               (culmen_depth_mm_0.0)            0.357357   
12                                    (species_Gentoo)            0.327327   
..                                                 ...                 ...   
617  (island_Biscoe, species_Gentoo, flipper_length...            0.159159   
618   (island_Biscoe, species_Gentoo, body_mass_g_2.0)            0.156156   
619  (culmen_depth_mm_0.0, flipper_length_mm_2.0, s...            0.201201   
620  (culmen_depth_mm_0.0, island_Biscoe, flipper_l...            0.201201   
621  (flipper_length_mm_2.0, island_Biscoe, species...            0.162162   

     consequent support   support  confidence      lift  representativity  \
0              0.489489  0.357357    1.000000  2.042945               1.0   
4              0.369369  0.204204    1.000000  2.707317               1.0   
8              0.438438  0.333333    1.000000  2.280822               1.0   
11             0.327327  0.312312    0.873950  2.669956               1.0   
12             0.357357  0.312312    0.954128  2.669956               1.0   
..                  ...       ...         ...       ...               ...   
617            0.330330  0.153153    0.962264  2.913036               1.0   
618            0.309309  0.153153    0.980769  3.170836               1.0   
619            0.252252  0.153153    0.761194  3.017591               1.0   
620            0.252252  0.153153    0.761194  3.017591               1.0   
621            0.297297  0.153153    0.944444  3.176768               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0    0.182435         inf       0.794393  0.730061   1.000000    0.865031  
4    0.128777         inf       0.792453  0.552846   1.000000    0.776423  
8    0.187187         inf       0.842342  0.760274   1.000000    0.880137  
11   0.195339    5.336537       0.973266  0.838710   0.812613    0.914039  
12   0.195339   14.009610       0.929816  0.838710   0.928620    0.914039  
..        ...         ...            ...       ...        ...         ...  
617  0.100578   17.746246       0.781022  0.455357   0.943650    0.712950  
618  0.104853   35.915916       0.811318  0.490385   0.972157    0.737957  
619  0.102400    3.131194       0.837019  0.510000   0.680633    0.684168  
620  0.102400    3.131194       0.837019  0.510000   0.680633    0.684168  
621  0.104943   12.648649       0.817837  0.500000   0.920940    0.729798  

[568 rows x 14 columns]

f) Para analizar las reglas vamos a ordenarlas por lift.

En la tabla se pueden obtener muchos patrones a partir de las reglas, entre ellos la relacion entre especie y isla que habita, por ejemplo, los pinguinos de especie Gentoo habitan mucho la isla Biscoe. Luego los machos suelen tener mayor altura y peso. Estos son algunos ejemplos, luego se puede analizar en profundidad para obtener mas patrones.

In [ ]:
pd.set_option("display.max_colwidth", None)

rules_filtradas_ordenadas = rules_filtradas.sort_values(by="lift", ascending=False)
rules_filtradas_ordenadas[["antecedents","consequents","support","confidence","lift","conviction"]].head(30)



antecedents  \
592                                                     (sex_MALE, species_Gentoo)   
570  (culmen_length_mm_2.0, island_Biscoe, flipper_length_mm_2.0, body_mass_g_2.0)   
420                                                     (sex_MALE, species_Gentoo)   
413                         (island_Biscoe, culmen_length_mm_2.0, body_mass_g_2.0)   
584                        (culmen_length_mm_2.0, species_Gentoo, body_mass_g_2.0)   
579                         (island_Biscoe, culmen_length_mm_2.0, body_mass_g_2.0)   
585                              (sex_MALE, species_Gentoo, flipper_length_mm_2.0)   
580                               (island_Biscoe, sex_MALE, flipper_length_mm_2.0)   
577                                      (sex_MALE, island_Biscoe, species_Gentoo)   
543                                                     (sex_MALE, species_Gentoo)   
403                                                     (sex_MALE, species_Gentoo)   
583                                    (sex_MALE, species_Gentoo, body_mass_g_2.0)   
582                   (culmen_length_mm_2.0, island_Biscoe, flipper_length_mm_2.0)   
398                   (culmen_length_mm_2.0, island_Biscoe, flipper_length_mm_2.0)   
590                 (culmen_length_mm_2.0, flipper_length_mm_2.0, body_mass_g_2.0)   
541                 (culmen_length_mm_2.0, flipper_length_mm_2.0, body_mass_g_2.0)   
415                                    (sex_MALE, species_Gentoo, body_mass_g_2.0)   
183                                          (island_Biscoe, culmen_length_mm_2.0)   
419                                          (island_Biscoe, culmen_length_mm_2.0)   
181                                                     (species_Gentoo, sex_MALE)   
572             (sex_MALE, flipper_length_mm_2.0, species_Gentoo, body_mass_g_2.0)   
568              (flipper_length_mm_2.0, island_Biscoe, sex_MALE, body_mass_g_2.0)   
399                              (sex_MALE, species_Gentoo, flipper_length_mm_2.0)   
396                               (island_Biscoe, sex_MALE, flipper_length_mm_2.0)   
410                                      (sex_MALE, island_Biscoe, species_Gentoo)   
327                                                     (species_Gentoo, sex_MALE)   
565               (sex_MALE, island_Biscoe, species_Gentoo, flipper_length_mm_2.0)   
536                              (sex_MALE, species_Gentoo, flipper_length_mm_2.0)   
442                               (island_Biscoe, sex_MALE, flipper_length_mm_2.0)   
445                             (flipper_length_mm_2.0, sex_MALE, body_mass_g_2.0)   

                                                                       consequents  \
592  (culmen_length_mm_2.0, island_Biscoe, flipper_length_mm_2.0, body_mass_g_2.0)   
570                                                     (sex_MALE, species_Gentoo)   
420                         (island_Biscoe, culmen_length_mm_2.0, body_mass_g_2.0)   
413                                                     (sex_MALE, species_Gentoo)   
584                               (island_Biscoe, sex_MALE, flipper_length_mm_2.0)   
579                              (sex_MALE, species_Gentoo, flipper_length_mm_2.0)   
585                         (island_Biscoe, culmen_length_mm_2.0, body_mass_g_2.0)   
580                        (culmen_length_mm_2.0, species_Gentoo, body_mass_g_2.0)   
577                 (culmen_length_mm_2.0, flipper_length_mm_2.0, body_mass_g_2.0)   
543                 (culmen_length_mm_2.0, flipper_length_mm_2.0, body_mass_g_2.0)   
403                   (culmen_length_mm_2.0, island_Biscoe, flipper_length_mm_2.0)   
583                   (culmen_length_mm_2.0, island_Biscoe, flipper_length_mm_2.0)   
582                                    (sex_MALE, species_Gentoo, body_mass_g_2.0)   
398                                                     (sex_MALE, species_Gentoo)   
590                                      (sex_MALE, island_Biscoe, species_Gentoo)   
541                         

g) En el futuro se puede, utilizar esta informacion para en caso de buscar un pinguino de la especie Gentoo ir a la isla Biscoe. Tambien pueden estudiar si por algun motivo especial en esa isla, los machos son mas grandes. O pueden predecir en el futuro las medidas de un pinguino de esa especie.